rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,133 @@
+| 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |   🔒   |
+| 1 _admin_, _SSO admin_ and _regular users_ | _SSO admin_ removes the `editor` role from _admin_ and SSO connector. _SSO admin_ logs out. |   🔒   |
+| _SSO admin_ and _regular users_            | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_      |   🔒   |
+
+Cases handled currently:
+
+| initial state               | operation                                                                          | result |
+| --------------------------- | ---------------------------------------------------------------------------------- | :----: |
+| _admin_ and _regular users_ | _admin_ tries to delete his account. Operation is not permitted, an error is shown |   🔓   |
+
+## Terminology
+
+- _admin_: local user with `editor` role
+- _SSO admin_: user with `editor` role added via SSO connector (eg. GitHub)
+- _regular user_: user without `editor` role
+- 🔒: No one can edit roles. Deadlock
+- 🔓: There is still an active user with an editor role
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows to addition SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can be `editors`.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least 2 local _admins_
+
+This will ensure that:
+
+- When one of the _admins_ loses access to the account the second one will operate.
+- Users won't be able to delete _admin_ user if there are only 2 _admins_. They would have to add a third one and then they can delete it.
+- Users won't be able to unassign the `editor` role from _admin_ user if there are only 2 _admins_. They would have to add a third and then they can unassign.
+- Role system lock will be independent of connector removal or SSO changes: team deletion/user deletion.
+
+Should we enforce this rule when a user is setting up its cluster? What about existing users? I think there are two approaches here:
+
+- they would have to create immediately a second _admin_ user (if they have only one currently). They cannot use a cluster if they don't.
```

Comment:
I don't think this is a viable option. Especially for trials or POCs, there might only be a single user interacting with the system. This would also impact any OSS users that might run Teleport for their personal access needs.

Reply:
That's fair. I agree, I think we should explore Naji's proposed approach (https://github.com/gravitational/teleport/pull/15706/files#r953966589) instead of mandating 2 admin users.